File: apps/web/src/lib/api/insights.ts
/**
 * Insights API client - M3 Sprint 6.
 *
 * Read-only surface for worker-generated analytics insights. The backend
 * storage/API names are intentionally kept stable; frontend copy remains
 * neutral and non-gamified.
 */

import { api } from './client';

// Generated OpenAPI shapes (issue #778) — re-exported so the contract drift
// guard in ./contracts.generated.ts is part of the real import graph.
export type { ApiInsightResponse, ApiInsightListResponse } from './contracts.generated';

export type InsightType =
  | 'pointbiserial'
  | 'null_association'
  | 'spearman'
  | 'weekday_pattern'
  | 'work_context_pattern'
  | 'weekday_context_pattern'
  | (string & {});
export type InsightTier = 'none' | 'early' | 'preliminary' | 'developing' | 'robust';
export type InsightMaturityPhase = 'collecting' | 'early_patterns' | 'provisional' | 'robust';

export interface InsightMaturity {
  phase: InsightMaturityPhase;
  phase_index: 1 | 2 | 3 | 4;
  current_entries: number;
  next_phase_at: number | null;
  next_phase_label: string | null;
  entries_until_next: number | null;
  user_message_key: string;
}

export interface InsightResponse {
  id: string;
  user_id: string;
  insight_type: InsightType;
  tier: InsightTier;
  metric: string;
  subject_type: string | null;
  subject_id: string | null;
  subject_label: string | null;
  effect_size: number | null;
  confidence: number | null;
  sample_n: number;
  statement: string | null;
  flags: Record<string, unknown>;
  payload: Record<string, unknown>;
  generated_for_date: string;
  generated_at: string;
  created_at: string;
  updated_at: string;
}

export interface InsightListResponse {
  insight_maturity: InsightMaturity;
  insights: InsightResponse[];
  /**
   * Timestamp of the latest completed generation attempt for this user.
   * This can be newer than every persisted row when a run correctly found no
   * candidates, so freshness must prefer it over `generated_at`.
   */
  last_successful_insight_run_at?: string | null;
  /** Latest finished USER_INSIGHTS attempt (Home compact worker status). */
  last_insight_run?: InsightWorkerRunSummary | null;
}

export interface InsightWorkerRunSummary {
  status: 'succeeded' | 'failed' | 'never_run';
  finished_at: string | null;
  started_at: string | null;
  insight_count: number | null;
  trigger_source: string | null;
  generated_for_date: string | null;
}

export interface InsightDigestItem {
  id: string;
  insight_type: InsightType;
  metric: string;
  effect_size: number | null;
  confidence: number | null;
  statement: string | null;
}

export interface InsightDigestResponse {
  week_start: string;
  week_end: string;
  insight_count: number;
  insights: InsightDigestItem[];
  push_title: string;
  push_body: string;
  // #739: set for a persisted weekly digest, null for the on-demand recompute
  // fallback. Only a stored digest (stable timestamp) drives the one-time modal.
  generated_at?: string | null;
}

export interface InsightListQuery {
  limit?: number;
}

/**
 * `/insights/latest` only. The chronological `/insights` endpoint declares no
 * family filter, so a shared query type would advertise — and silently drop —
 * an option that endpoint ignores.
 */
export interface LatestInsightListQuery extends InsightListQuery {
  /**
   * Restrict to these insight families *before* the row cap applies. A surface
   * that renders one or two families otherwise loses valid rows to unrelated
   * subjects occupying the first `limit` slots (#959).
   */
  insightTypes?: readonly string[];
  /** Restrict before the row cap to insights containing every structured signal. */
  pairSignals?: readonly { kind: string; id: string }[];
}

export type TagCooccurrenceRange = '7d' | '30d' | '90d' | '1y';

export interface TagCooccurrenceTagRef {
  tag_id: string;
  slug: string;
  name: string;
  category: string;
  color: string | null;
}

export interface TagCooccurrencePair {
  tag_a: TagCooccurrenceTagRef;
  tag_b: TagCooccurrenceTagRef;
  count: number;
  pct_of_a: number;
  pct_of_b: number;
}

export interface TagCooccurrenceResponse {
  range: TagCooccurrenceRange;
  start_date: string;
  end_date: string;
  min_count: number;
  pairs: TagCooccurrencePair[];
  /**
   * The window holds fewer logged days than the analysis needs. An empty
   * `pairs` then means "cannot be computed here", not "nothing found" — a 7-day
   * range can never reach the floor (#966).
   */
  window_too_short?: boolean;
}

export interface TagCooccurrenceQuery {
  range?: TagCooccurrenceRange;
  min_count?: number;
}

export interface TagClusterMember {
  kind: 'tag' | 'symptom';
  signal_id: string;
  slug: string;
  name: string;
  icon?: string | null;
  category?: string | null;
  color?: string | null;
}

export interface TagClusterGroup {
  cluster_id: number;
  label: string;
  tags: TagCooccurrenceTagRef[];
  members: TagClusterMember[];
  cluster_kind: 'tags_only' | 'mixed';
  strength: number;
}

export interface TagClustersResponse {
  status: 'ok' | 'insufficient_data';
  entry_count: number;
  active_tag_count: number;
  active_signal_count: number;
  window_days: number;
  k: number | null;
  reason: string | null;
  cluster_kind: 'tags_only' | 'mixed';
  cluster_maturity?: 'early' | 'provisional' | 'robust' | null;
  cluster_mode?: 'pair' | 'kmeans' | null;
  entries_until_robust?: number | null;
  silhouette_score?: number | null;
  clusters: TagClusterGroup[];
  /** #706: transparency + client-side strength bands (defaulted for old backends). */
  shown_cluster_count?: number;
  omitted_signal_count?: number;
  strength_floor?: number;
}

export interface InsightRegenerateResponse {
  status: 'ok';
  generated_for_date: string;
  insight_count: number;
  tag_clusters_status: 'ok' | 'insufficient_data';
  trigger_source: string;
}

export interface SymptomTagCooccurrenceSymptomRef {
  symptom_id: string;
  slug: string;
  name: string;
  icon: string | null;
}

export type SymptomTagCooccurrenceConfounder = 'weekday' | 'work_context' | 'calendar_context';

export interface SymptomTagCooccurrenceCell {
  symptom: SymptomTagCooccurrenceSymptomRef;
  tag: TagCooccurrenceTagRef;
  phi: number;
  jaccard: number;
  lift: number;
  co_count: number;
  symptom_count: number;
  tag_count: number;
  total_count: number;
  p_value_corrected: number;
  confounder: SymptomTagCooccurrenceConfounder | null;
}

export interface SymptomTagCooccurrenceResponse {
  range: TagCooccurrenceRange;
  start_date: string;
  end_date: string;
  min_count: number;
  cells: SymptomTagCooccurrenceCell[];
}

function buildQuery(query: InsightListQuery): string {
  const params = new URLSearchParams();
  if (query.limit !== undefined) params.set('limit', String(query.limit));
  const qs = params.toString();
  return qs ? `?${qs}` : '';
}

/**
 * Separate builder rather than a branch in `buildQuery`: `/insights` physically
 * cannot emit `insight_type` this way, instead of merely being typed not to.
 */
function buildLatestQuery(query: LatestInsightListQuery): string {
  const params = new URLSearchParams();
  if (query.limit !== undefined) params.set('limit', String(query.limit));
  // Repeated `insight_type=` params — FastAPI reads them as a list.
  for (const type of query.insightTypes ?? []) params.append('insight_type', type);
  for (const signal of query.pairSignals ?? []) {
    params.append('pair_signal', `${signal.kind}:${signal.id}`);
  }
  const qs = params.toString();
  return qs ? `?${qs}` : '';
}

/** GET /insights/latest - list latest generated insights by analytical subject. */
export async function listLatestInsights(
  query: LatestInsightListQuery = {}
): Promise<InsightListResponse> {
  return api.get<InsightListResponse>(`/insights/latest${buildLatestQuery(query)}`);
}

/** GET /insights - chronological insight history (newest-first, no subject dedupe). */
export async function listInsights(query: InsightListQuery = {}): Promise<InsightListResponse> {
  return api.get<InsightListResponse>(`/insights${buildQuery(query)}`);
}

/** GET /insights/tag-cooccurrence - tag pair counts for the co-occurrence heatmap (M5.1). */
export async function fetchTagCooccurrence(
  query: TagCooccurrenceQuery = {}
): Promise<TagCooccurrenceResponse> {
  const params = new URLSearchParams();
  if (query.range) params.set('range', query.range);
  if (query.min_count !== undefined) params.set('min_count', String(query.min_count));
  const qs = params.toString();
  return api.get<TagCooccurrenceResponse>(
    qs ? `/insights/tag-cooccurrence?${qs}` : '/insights/tag-cooccurrence'
  );
}

/** GET /insights/tag-clusters - M7 tag groups that often appear together. */
export async function fetchTagClusters(): Promise<TagClustersResponse> {
  return api.get<TagClustersResponse>('/insights/tag-clusters');
}

/** GET /insights/symptom-tag-cooccurrence - symptom x tag lift cells for M7. */
export async function fetchSymptomTagCooccurrence(
  query: TagCooccurrenceQuery = {}
): Promise<SymptomTagCooccurrenceResponse> {
  const params = new URLSearchParams();
  if (query.range) params.set('range', query.range);
  if (query.min_count !== undefined) params.set('min_count', String(query.min_count));
  const qs = params.toString();
  return api.get<SymptomTagCooccurrenceResponse>(
    qs ? `/insights/symptom-tag-cooccurrence?${qs}` : '/insights/symptom-tag-cooccurrence'
  );
}

export interface InsightEventWindowResponse {
  onset: string;
  label: string | null;
}

export interface InsightEventWindowsResponse {
  range: TagCooccurrenceRange;
  start_date: string;
  end_date: string;
  events: InsightEventWindowResponse[];
  points: import('./stats').TimeseriesPoint[];
  /** #488: lag insights align on the feature; the outcome is expected at +lag_days. */
  lag_days?: number | null;
}

/** POST /insights/regenerate — on-demand insight + tag-cluster regeneration (M10.1). */
export async function regenerateInsights(): Promise<InsightRegenerateResponse> {
  return api.post<InsightRegenerateResponse>('/insights/regenerate');
}

/** GET /insights/{id}/event-windows — ADR-0035 §6 explore-events data. */
export async function fetchInsightEventWindows(
  insightId: string,
  range: TagCooccurrenceRange,
  options?: { signal?: AbortSignal }
): Promise<InsightEventWindowsResponse> {
  const params = new URLSearchParams({ range });
  const path = `/insights/${encodeURIComponent(insightId)}/event-windows?${params}`;
  return options
    ? api.get<InsightEventWindowsResponse>(path, options)
    : api.get<InsightEventWindowsResponse>(path);
}

/** GET /insights/{id} — single insight for Layer-2 signal detail. */
export async function fetchInsight(
  insightId: string,
  options?: { signal?: AbortSignal }
): Promise<InsightResponse> {
  const path = `/insights/${encodeURIComponent(insightId)}`;
  return options ? api.get<InsightResponse>(path, options) : api.get<InsightResponse>(path);
}

export interface InsightVerificationPoint {
  date: string;
  value: number;
  present: boolean;
}

export interface InsightVerificationResponse {
  range: TagCooccurrenceRange;
  start_date: string;
  end_date: string;
  metric: string;
  subject_label: string | null;
  points: InsightVerificationPoint[];
  with_mean: number | null;
  without_mean: number | null;
  with_se: number | null;
  without_se: number | null;
  with_n: number;
  without_n: number;
}

/** GET /insights/{id}/verification — with/without day series (Phase 7 / G1). */
export async function fetchInsightVerification(
  insightId: string,
  range: TagCooccurrenceRange = '90d',
  options?: { signal?: AbortSignal }
): Promise<InsightVerificationResponse> {
  const params = new URLSearchParams({ range });
  const path = `/insights/${encodeURIComponent(insightId)}/verification?${params}`;
  return options
    ? api.get<InsightVerificationResponse>(path, options)
    : api.get<InsightVerificationResponse>(path);
}

export async function fetchLatestInsightDigest(): Promise<InsightDigestResponse> {
  return api.get<InsightDigestResponse>('/insights/digest/latest');
}

export type InsightHistoryVisibility = 'active' | 'dismissed' | 'all';

export interface InsightHistoryItem extends InsightResponse {
  visibility: 'active' | 'dismissed';
  subject_key: string;
  first_seen_on: string | null;
  last_seen_on: string | null;
  observation_count: number | null;
}

export interface InsightHistoryResponse {
  insights: InsightHistoryItem[];
  total: number;
  limit: number;
  offset: number;
}

export interface InsightHistoryQuery {
  status?: InsightHistoryVisibility;
  from?: string;
  to?: string;
  limit?: number;
  offset?: number;
}

/** GET /insights/history — chronological timeline / archive (#601 Phase 2). */
export async function listInsightHistory(
  query: InsightHistoryQuery = {}
): Promise<InsightHistoryResponse> {
  const params = new URLSearchParams();
  if (query.status) params.set('status', query.status);
  if (query.from) params.set('from', query.from);
  if (query.to) params.set('to', query.to);
  if (query.limit !== undefined) params.set('limit', String(query.limit));
  if (query.offset !== undefined) params.set('offset', String(query.offset));
  const qs = params.toString();
  return api.get<InsightHistoryResponse>(qs ? `/insights/history?${qs}` : '/insights/history');
}

export interface InsightDismissalResponse {
  id: string;
  subject_key: string;
  insight_id: string | null;
  dismissed_at: string;
  created_at: string;
  insight: InsightResponse | null;
}

export interface InsightDismissalListResponse {
  dismissals: InsightDismissalResponse[];
}

/** GET /insights/dismissals — subject-stable hidden insights. */
export async function listInsightDismissals(): Promise<InsightDismissalListResponse> {
  return api.get<InsightDismissalListResponse>('/insights/dismissals');
}

/** POST /insights/dismissals — hide by insight id (subject-stable). */
export async function createInsightDismissal(insightId: string): Promise<InsightDismissalResponse> {
  return api.post<InsightDismissalResponse>('/insights/dismissals', {
    insight_id: insightId,
  });
}

/** DELETE /insights/dismissals/{id} — undo hide. */
export async function deleteInsightDismissal(dismissalId: string): Promise<void> {
  await api.delete(`/insights/dismissals/${encodeURIComponent(dismissalId)}`);
}

/** DELETE /insights/dismissals/by-insight/{id} — undo by insight id. */
export async function deleteInsightDismissalByInsightId(insightId: string): Promise<void> {
  await api.delete(`/insights/dismissals/by-insight/${encodeURIComponent(insightId)}`);
}

File: backend/app/api/v1/endpoints/insights.py
"""Insight read and regeneration endpoints (M3, M10.1)."""

from __future__ import annotations

import uuid
from datetime import date

import redis.asyncio as aioredis
from fastapi import APIRouter, Depends, HTTPException, Query, Request, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.v1.deps.auth import (
    get_current_insight_trigger_admin,
    get_current_verified_user,
)
from app.core.rate_limit import limiter
from app.db.redis_client import get_redis
from app.db.session import get_session
from app.models.insight import Insight, InsightType
from app.models.insight_dismissal import InsightDismissal
from app.models.user import User
from app.schemas.insight import (
    InsightDigestItemResponse,
    InsightDigestResponse,
    InsightDismissalCreate,
    InsightDismissalListResponse,
    InsightDismissalResponse,
    InsightEventWindowsResponse,
    InsightHistoryItem,
    InsightHistoryResponse,
    InsightListResponse,
    InsightRegenerateResponse,
    InsightResponse,
    InsightTriggerResponse,
)
from app.schemas.stats import (
    SymptomTagCooccurrenceResponse,
    TagClustersResponse,
    TagCooccurrenceRange,
    TagCooccurrenceResponse,
)
from app.services.insight_digest import (
    DigestDisabledError,
    DigestNotAvailableError,
    build_push_payload,
    get_latest_weekly_digest,
)
from app.services.insight_dismissal_service import (
    InsightDismissalNotFoundError,
    InsightDismissalView,
    create_insight_dismissal,
    delete_insight_dismissal,
    delete_insight_dismissal_by_insight_id,
    list_insight_dismissals,
)
from app.services.insight_service import (
    DEFAULT_INSIGHT_LIST_LIMIT,
    DEFAULT_LATEST_INSIGHT_LIMIT,
    MAX_INSIGHT_LIST_LIMIT,
    MAX_LATEST_INSIGHT_LIMIT,
    InsightEventWindowsUnsupportedError,
    InsightNotFoundError,
    get_insight_event_windows,
    get_insight_maturity,
    list_insight_history,
    list_insights,
    list_latest_insights,
)
from app.services.insight_worker_service import (
    AnalyticsDisabledError,
    InsightJobNotFoundError,
    regenerate_insights_for_user,
    try_acquire_regenerate_slot,
)
from app.services.stats_service import get_symptom_tag_cooccurrence, get_tag_cooccurrence
from app.services.tag_cluster_service import get_tag_clusters
from app.workers.analytics import run_insights_once

router = APIRouter()


def _cooccurrence_range_query(
    range: str = Query(default="90d", alias="range"),
) -> TagCooccurrenceRange:
    if range not in {"7d", "30d", "90d", "1y"}:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="range must be one of 7d, 30d, 90d, 1y",
        )
    return range  # type: ignore[return-value]


@router.get(
    "/digest/latest",
    response_model=InsightDigestResponse,
    summary="Latest weekly insight digest",
)
@limiter.limit("60/minute")
async def get_latest_digest_endpoint(
    request: Request,
    user: User = Depends(get_current_verified_user),
    db: AsyncSession = Depends(get_session),
) -> InsightDigestResponse:
    try:
        digest = await get_latest_weekly_digest(db, user_id=user.id)
    except DigestDisabledError as exc:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Weekly insight digest is disabled for this account",
        ) from exc
    except DigestNotAvailableError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Not enough recent insights for a weekly digest",
        ) from exc

    push = build_push_payload(digest)
    return InsightDigestResponse(
        week_start=digest.week_start,
        week_end=digest.week_end,
        insight_count=digest.insight_count,
        insights=[
            InsightDigestItemResponse(
                id=item.id,
                insight_type=(
                    item.insight_type
                    if isinstance(item.insight_type, InsightType)
                    else InsightType(item.insight_type)
                ),
                metric=item.metric,
                effect_size=item.effect_size,
                confidence=item.confidence,
                statement=item.statement,
            )
            for item in digest.insights
        ],
        push_title=push["title"],
        push_body=push["body"],
    )


@router.get(
    "",
    response_model=InsightListResponse,
    summary="List generated insights",
)
@limiter.limit("120/minute")
async def list_insights_endpoint(
    request: Request,
    limit: int = Query(default=DEFAULT_INSIGHT_LIST_LIMIT, ge=1, le=MAX_INSIGHT_LIST_LIMIT),
    user: User = Depends(get_current_verified_user),
    db: AsyncSession = Depends(get_session),
) -> InsightListResponse:
    insights = await list_insights(db, user_id=user.id, limit=limit)
    insight_maturity = await get_insight_maturity(db, user_id=user.id)
    return InsightListResponse(
        insight_maturity=insight_maturity,
        insights=[InsightResponse.model_validate(insight) for insight in insights],
    )


@router.get(
    "/latest",
    response_model=InsightListResponse,
    summary="List latest generated insights by analytical subject",
)
@limiter.limit("120/minute")
async def list_latest_insights_endpoint(
    request: Request,
    limit: int = Query(default=DEFAULT_LATEST_INSIGHT_LIMIT, ge=1, le=MAX_LATEST_INSIGHT_LIMIT),
    user: User = Depends(get_current_verified_user),
    db: AsyncSession = Depends(get_session),
) -> InsightListResponse:
    insights = await list_latest_insights(db, user_id=user.id, limit=limit)
    insight_maturity = await get_insight_maturity(db, user_id=user.id)
    return InsightListResponse(
        insight_maturity=insight_maturity,
        insights=[InsightResponse.model_validate(insight) for insight in insights],
    )


@router.get(
    "/history",
    response_model=InsightHistoryResponse,
    summary="Chronological insight history for timeline / archive",
)
@limiter.limit("60/minute")
async def list_insight_history_endpoint(
    request: Request,
    status_filter: str = Query(default="all", alias="status"),
    from_date: date | None = Query(default=None, alias="from"),
    to_date: date | None = Query(default=None, alias="to"),
    limit: int = Query(default=DEFAULT_INSIGHT_LIST_LIMIT, ge=1, le=MAX_INSIGHT_LIST_LIMIT),
    offset: int = Query(default=0, ge=0),
    user: User = Depends(get_current_verified_user),
    db: AsyncSession = Depends(get_session),
) -> InsightHistoryResponse:
    if status_filter not in {"active", "dismissed", "all"}:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="status must be one of active, dismissed, all",
        )
    entries, total = await list_insight_history(
        db,
        user_id=user.id,
        status=status_filter,
        from_date=from_date,
        to_date=to_date,
        limit=limit,
        offset=offset,
    )
    return InsightHistoryResponse(
        insights=[
            InsightHistoryItem(
                **InsightResponse.model_validate(entry.insight).model_dump(),
                visibility=entry.visibility,  # type: ignore[arg-type]
                subject_key=entry.subject_key,
                first_seen_on=entry.first_seen_on,
                last_seen_on=entry.last_seen_on,
                observation_count=entry.observation_count,
            )
            for entry in entries
        ],
        total=total,
        limit=limit,
        offset=offset,
    )


@router.get(
    "/tag-cooccurrence",
    response_model=TagCooccurrenceResponse,
    summary="Tag co-occurrence pairs for heatmap visualisation",
)
@limiter.limit("120/minute")
async def get_tag_cooccurrence_endpoint(
    request: Request,
    range: TagCooccurrenceRange = Depends(_cooccurrence_range_query),
    min_count: int = Query(default=2, ge=1, le=100),
    user: User = Depends(get_current_verified_user),
    db: AsyncSession = Depends(get_session),
) -> TagCooccurrenceResponse:
    return await get_tag_cooccurrence(
        db,
        user_id=user.id,
        range_=range,
        min_count=min_count,
    )


@router.get(
    "/symptom-tag-cooccurrence",
    response_model=SymptomTagCooccurrenceResponse,
    summary="Symptom x tag co-occurrence cells for heatmap visualisation",
)
@limiter.limit("120/minute")
async def get_symptom_tag_cooccurrence_endpoint(
    request: Request,
    range: TagCooccurrenceRange = Depends(_cooccurrence_range_query),
    min_count: int = Query(default=3, ge=1, le=100),
    user: User = Depends(get_current_verified_user),
    db: AsyncSession = Depends(get_session),
) -> SymptomTagCooccurrenceResponse:
    return await get_symptom_tag_cooccurrence(
        db,
        user_id=user.id,
        range_=range,
        min_count=min_count,
    )


@router.get(
    "/tag-clusters",
    response_model=TagClustersResponse,
    summary="Tag groups that often appear together",
)
@limiter.limit("120/minute")
async def get_tag_clusters_endpoint(
    request: Request,
    user: User = Depends(get_current_verified_user),
    db: AsyncSession = Depends(get_session),
) -> TagClustersResponse:
    return await get_tag_clusters(db, user_id=user.id)


def _dismissal_response(
    view_or_row: InsightDismissalView | InsightDismissal,
    insight: Insight | None = None,
) -> InsightDismissalResponse:
    if isinstance(view_or_row, InsightDismissalView):
        row = view_or_row.dismissal
        insight = view_or_row.insight
    else:
        row = view_or_row
    return InsightDismissalResponse(
        id=row.id,
        subject_key=row.subject_key,
        insight_id=row.insight_id,
        dismissed_at=row.dismissed_at,
        created_at=row.created_at,
        insight=InsightResponse.model_validate(insight) if insight is not None else None,
    )


@router.get(
    "/dismissals",
    response_model=InsightDismissalListResponse,
    summary="List subject-stable hidden insights",
)
@limiter.limit("60/minute")
async def list_insight_dismissals_endpoint(
    request: Request,
    user: User = Depends(get_current_verified_user),
    db: AsyncSession = Depends(get_session),
) -> InsightDismissalListResponse:
    views = await list_insight_dismissals(db, user_id=user.id)
    return InsightDismissalListResponse(dismissals=[_dismissal_response(view) for view in views])


@router.post(
    "/dismissals",
    response_model=InsightDismissalResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Hide an insight by subject-stable key",
)
@limiter.limit("60/minute")
async def create_insight_dismissal_endpoint(
    request: Request,
    payload: InsightDismissalCreate,
    user: User = Depends(get_current_verified_user),
    db: AsyncSession = Depends(get_session),
) -> InsightDismissalResponse:
    try:
        row = await create_insight_dismissal(db, user_id=user.id, insight_id=payload.insight_id)
    except InsightNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Insight not found",
        ) from exc
    await db.refresh(row)
    return _dismissal_response(row)


@router.delete(
    "/dismissals/by-insight/{insight_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Undo hide for the subject of an insight id",
)
@limiter.limit("60/minute")
async def delete_insight_dismissal_by_insight_endpoint(
    request: Request,
    insight_id: uuid.UUID,
    user: User = Depends(get_current_verified_user),
    db: AsyncSession = Depends(get_session),
) -> None:
    await delete_insight_dismissal_by_insight_id(db, user_id=user.id, insight_id=insight_id)


@router.delete(
    "/dismissals/{dismissal_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Undo a subject-stable insight hide",
)
@limiter.limit("60/minute")
async def delete_insight_dismissal_endpoint(
    request: Request,
    dismissal_id: uuid.UUID,
    user: User = Depends(get_current_verified_user),
    db: AsyncSession = Depends(get_session),
) -> None:
    try:
        await delete_insight_dismissal(db, user_id=user.id, dismissal_id=dismissal_id)
    except InsightDismissalNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Insight dismissal not found",
        ) from exc


@router.post(
    "/regenerate",
    response_model=InsightRegenerateResponse,
    summary="Regenerate insights and tag clusters for the current user",
)
@limiter.limit("10/minute")
async def regenerate_insights_endpoint(
    request: Request,
    user: User = Depends(get_current_verified_user),
    db: AsyncSession = Depends(get_session),
    redis: aioredis.Redis = Depends(get_redis),
) -> InsightRegenerateResponse:
    if not await try_acquire_regenerate_slot(user_id=user.id):
        raise HTTPException(
            status_code=status.HTTP_429_TOO_MANY_REQUESTS,
            detail="Insight regeneration is limited to once per hour",
        )
    try:
        result = await regenerate_insights_for_user(
            db,
            user_id=user.id,
            trigger_source="user_regenerate",
        )
        await db.commit()
    except AnalyticsDisabledError as exc:
        await db.rollback()
        await redis.delete(f"insight:regenerate:{user.id}")
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Analytics processing is disabled for this account",
        ) from exc
    except InsightJobNotFoundError as exc:
        await db.rollback()
        await redis.delete(f"insight:regenerate:{user.id}")
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="No encryption key found for insight generation",
        ) from exc

    return InsightRegenerateResponse(
        generated_for_date=result.generated_for_date,
        insight_count=result.insight_count,
        tag_clusters_status=result.tag_clusters_status,
        trigger_source=result.trigger_source,
    )


@router.post(
    "/trigger",
    response_model=InsightTriggerResponse,
    summary="Run scheduled insight generation for all eligible users (admin)",
)
@limiter.limit("10/minute")
async def trigger_insights_endpoint(
    request: Request,
    _admin: User = Depends(get_current_insight_trigger_admin),
) -> InsightTriggerResponse:
    summary = await run_insights_once(trigger_source="admin_trigger")
    return InsightTriggerResponse(
        eligible_users=summary.eligible_users,
        processed_users=summary.processed_users,
        failed_users=summary.failed_users,
        generated_insights=summary.generated_insights,
    )


@router.get(
    "/{insight_id}/event-windows",
    response_model=InsightEventWindowsResponse,
    summary="Event-aligned onset dates and timeseries for an insight",
)
@limiter.limit("120/minute")
async def get_insight_event_windows_endpoint(
    request: Request,
    insight_id: uuid.UUID,
    range: TagCooccurrenceRange = Depends(_cooccurrence_range_query),
    user: User = Depends(get_current_verified_user),
    db: AsyncSession = Depends(get_session),
) -> InsightEventWindowsResponse:
    try:
        return await get_insight_event_windows(
            db,
            user_id=user.id,
            insight_id=insight_id,
            range_=range,
        )
    except InsightNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Insight not found",
        ) from exc
    except InsightEventWindowsUnsupportedError as exc:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="Event windows are only available for tag and symptom insights",
        ) from exc
